lightning/src/util/message_signing.rs

```diff
@@ -0,0 +1,145 @@
+//! from a specific node.
+//! Furthermore, this tools can be used to sign / verify messages using ephemeral keys not tight to node's identities.
+//!
+//! Note this is not parts of the specs, but follows lnd's signing and verifying protocol, which can is defined as follows:
+//!
+//! signature = zbase32(SigRec(sha256d(("Lightning Signed Message:" + msg)))
+//! zbase32 from https://philzimmermann.com/docs/human-oriented-base-32-encoding.txt
+//! SigRec has first byte 31 + recovery id, followed by 64 byte sig.
+//!
+//! This implementation is compatible with boths lnd's and c-lightning's
+//!
+//! https://lightning.readthedocs.io/lightning-signmessage.7.html
+//! https://api.lightning.community/#signmessage
+
+
+use crate::util::zbase32;
+use bitcoin::hashes::{sha256d, Hash};
+use bitcoin::secp256k1::recovery::{RecoverableSignature, RecoveryId};
+use bitcoin::secp256k1::{Error, Message, PublicKey, Secp256k1, SecretKey};
+
+static LN_MESSAGE_PREFIX: &[u8] = b"Lightning Signed Message:";
+
+
+fn sigrec_encode(sig_rec: RecoverableSignature) -> Vec<u8> {
+    let (rid, rsig) = sig_rec.serialize_compact();
+    let prefix = rid.to_i32() as u8 + 31;
+
+    [&[prefix], &rsig[..]].concat()
+}
+
+
+fn sigrec_decode(sig_rec: Vec<u8>) -> Result<RecoverableSignature, Error> {
+    let rsig = &sig_rec[1..];
+    let rid = sig_rec[0] as i32 - 31;
+
+    match RecoveryId::from_i32(rid) {
+        Ok(x) => RecoverableSignature::from_compact(rsig, x),
+        Err(e) => Err(e)
+    }
+}
+
+/// Creates a digital signature of a message given a SecretKey, like the node's secret.
+/// A receiver knowing the PublicKey (e.g. the node's id) and the message can be sure that the signature
+/// was generated by the claimer. Signatures are EC recoverable, meaning that given the message and the signature
+/// the PublicKey of the signer can be extracted.
+pub fn sign(msg: &[u8], sk: SecretKey) -> Result<String, Error> {
+    let secp_ctx = Secp256k1::signing_only();
+    let msg_hash = sha256d::Hash::hash(&[LN_MESSAGE_PREFIX, msg].concat());
+
+    let sig = secp_ctx.sign_recoverable(&Message::from_slice(&msg_hash)?, &sk);
+    Ok(zbase32::encode(&sigrec_encode(sig)))
+}
+
+
+/// Recovers the PublicKey of the signer of the message given the message and the signature.
+pub fn recover_pk(msg: &[u8], sig: &str) ->  Result<PublicKey, Error>{
+    let secp_ctx = Secp256k1::verification_only();
+    let msg_hash = sha256d::Hash::hash(&[LN_MESSAGE_PREFIX, msg].concat());
+
+    match zbase32::decode(&sig) {
+        Ok(sig_rec) => {
+            match sigrec_decode(sig_rec) {
+                Ok(sig) => secp_ctx.recover(&Message::from_slice(&msg_hash)?, &sig),
+                Err(e) => Err(e)
+            }
+        },
+        Err(_) => Err(Error::InvalidSignature)
+    }
+}
+
+/// Verifies a message was signed by a PrivateKey that derives to a given PublicKey, given a message, a signature,
+/// and the PublicKey.
+pub fn verify(msg: &[u8], sig: &str, pk: PublicKey) -> bool {
```

Comment:
I tested this with the sample and `lnd`'s telling me the message is invalid (the pubkey's correct): https://i.imgur.com/lizoFib.png

Haven't tested further. 
Instructions to reproduce/test further (I used Polar but feel free to use whatever you want):
* clone this branch of the sample: https://github.com/valentinewallace/ldk-sample/tree/for-testing-signing (it uses a branch of `rust-lightning` that's just your branch + one commit to expose a struct that's needed)
* I used Polar: https://github.com/jamaljsr/polar by cloning and running `yarn dev`
* start a new test network in Polar and add an `lnd` node
* start the sample: `cargo run polaruser:polarpass@127.0.0.1:18443 ./ 9738 regtest`
* in the sample CLI: `connectpeer <lnd node>`, where `lnd node`'s found under `alice > Connect > P2P External`
* in the sample CLI: `signmessage` (the message is hardcoded, see: https://github.com/valentinewallace/ldk-sample/blob/for-testing-signing/src/cli.rs#L378)
* copy this output, then open a terminal for the lnd node under `alice > Actions > Terminal > Launch`
* `lncli verifymessage --msg abcdefghijklmnopqrstuvwxyz1234567890 --sig <sig you just copied>`

Also added `verifymessage` to the sample: https://github.com/valentinewallace/ldk-sample/blob/for-testing-signing/src/cli.rs#L390 if you wanna test that

Reply:
> I tested this with the sample and lnd's telling me the message is invalid (the pubkey's correct): https://i.imgur.com/lizoFib.png

Wouldn't this mean the two nodes are not properly connected (or the node that originated the signature is not on the verifiers database?). I'm wondering since you point that the recovered public key is correct, so the signature is actually valid.

From [lnd docs](https://api.lightning.community/#verifymessage):

> VerifyMessage verifies a signature over a msg. The signature must be zbase32 encoded and signed by an active node in the resident node's channel database. In addition to returning the validity of the signature, VerifyMessage also returns the recovered pubkey from the signature.

Reply:
@valentinewallace I double checked this against `lnd` and `c-lightning`. Both return:

```
{
   "pubkey": our_public_key,
   "verified": false
}
```

However, in `c-lightning`, if our public key is passed as optional parameter then the message is verified.

Checking [c-lightning's docs](https://lightning.readthedocs.io/lightning-checkmessage.7.html), we can find the following:

> As a special case, if pubkey is not specified, we will try every known node key (as per listnodes), and verification succeeds if it matches for any one of them.

And, If I check `listnodes`, the list is empty (even though the node is one of our peers), so I'm guessing that's the root of the issue. The pubkey is correctly recovered in both cases, but the node seems not to be part of neither of the peers known nodes, so it fails validation agains known nodes.

Reply:
Furthermore, querying `describegraph` in lnd's node does not show our testing node, which I'd say backs the aforementioned claim.

I'm not sure why that's happening though. Any thoughts?

Reply:
Asked around #c-lightning on IRC, looks like a channel announced is required to show-up in `listnodes`, and I'm assuming the same may apply to `describegraph` in `lnd`:

```
20:12 <sr_gi> Hey guys, what is needed for a node to show in listnodes? I was doing some testing using Polar and
connecting a c-lightning node to a rust-lightning node and even though they were connected the peer did no show up
in the list. Is a channel required or something on those lines?
20:14 <@cdecker> An announced channel is required, yes
```

Reply:
As discussed offline, this seems like a reasonable explanation for why lnd's giving `verified: false` :)